File: python/fastapi-api/routers/public.py
"""Public read surface: articles, taxonomies, authors, companies, search."""

from typing import Annotated, Sequence, TypeVar

from fastapi import APIRouter, HTTPException, Path, Query, status

from dependencies import StoreDep
from models import (
    ArticleDetail,
    ArticlePage,
    AuthorProfile,
    CompanyPage,
    Lang,
    SearchResults,
    Stage,
    TaxonomyItem,
)

router = APIRouter(tags=["public"])

PageNumber = Annotated[int, Query(ge=1, description="1-based page number")]
PageLimit = Annotated[int, Query(ge=1, le=100, description="Items per page")]

T = TypeVar("T")


def _window(items: Sequence[T], page: int, limit: int) -> tuple[Sequence[T], int, int]:
    total = len(items)
    pages = -(-total // limit)
    start = (page - 1) * limit
    return items[start : start + limit], total, pages


@router.get("/articles", response_model=ArticlePage, summary="List articles")
async def list_articles(
    store: StoreDep,
    page: PageNumber = 1,
    limit: PageLimit = 20,
    category: Annotated[str | None, Query(description="Category slug")] = None,
    tag: Annotated[str | None, Query(description="Tag slug")] = None,
    author: Annotated[str | None, Query(description="Author slug")] = None,
    lang: Annotated[Lang | None, Query()] = None,
    q: Annotated[str | None, Query(description="Matches title and excerpt")] = None,
) -> ArticlePage:
    records = store.filter_articles(
        category=category, tag=tag, author=author, lang=lang, q=q
    )
    window, total, pages = _window(records, page, limit)
    return ArticlePage(
        items=[record.summary for record in window],
        page=page,
        limit=limit,
        total=total,
        pages=pages,
    )


@router.get(
    "/articles/{slug}",
    response_model=ArticleDetail,
    responses={404: {"description": "No article with that slug"}},
)
async def get_article(store: StoreDep, slug: Annotated[str, Path()]) -> ArticleDetail:
    record = store.by_slug.get(slug)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, f"No article with slug {slug!r}")
    return store.build_detail(record)


@router.get("/categories", response_model=list[TaxonomyItem])
async def list_categories(store: StoreDep) -> list[TaxonomyItem]:
    return store.category_listing()


@router.get("/tags", response_model=list[TaxonomyItem])
async def list_tags(store: StoreDep) -> list[TaxonomyItem]:
    return store.tag_listing()


@router.get(
    "/authors/{slug}",
    response_model=AuthorProfile,
    responses={404: {"description": "No author with that slug"}},
)
async def get_author(store: StoreDep, slug: Annotated[str, Path()]) -> AuthorProfile:
    profile = store.author_profile(slug)
    if profile is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, f"No author with slug {slug!r}")
    return profile


@router.get("/companies", response_model=CompanyPage)
async def list_companies(
    store: StoreDep,
    page: PageNumber = 1,
    limit: PageLimit = 20,
    industry: Annotated[str | None, Query()] = None,
    stage: Annotated[Stage | None, Query()] = None,
    min_funding: Annotated[int | None, Query(ge=0, description="USD")] = None,
) -> CompanyPage:
    companies = store.filter_companies(
        industry=industry, stage=stage, min_funding=min_funding
    )
    window, total, pages = _window(companies, page, limit)
    return CompanyPage(
        items=list(window), page=page, limit=limit, total=total, pages=pages
    )


@router.get("/search", response_model=SearchResults)
async def search(
    store: StoreDep,
    q: Annotated[str, Query(min_length=2, max_length=100)],
) -> SearchResults:
    articles, companies = store.search(q)
    return SearchResults(query=q, articles=articles, companies=companies)

File: rust/actix-api/src/routes/public.rs
//! Unauthenticated read surface.

use std::collections::HashMap;

use actix_web::{HttpResponse, get, web};
use serde::Deserialize;

use crate::dto::{self, AuthorDetail, Facet, Page, SearchResults};
use crate::error::ApiError;
use crate::routes::{offset, page_limit};
use crate::store::{AppState, Article};
use crate::validation::{LANGS, STAGES};

#[derive(Debug, Deserialize)]
pub struct ListQuery {
    page: Option<u32>,
    limit: Option<u32>,
    category: Option<String>,
    tag: Option<String>,
    author: Option<String>,
    lang: Option<String>,
    q: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CompanyQuery {
    page: Option<u32>,
    limit: Option<u32>,
    industry: Option<String>,
    stage: Option<String>,
    min_funding: Option<u64>,
}

#[derive(Debug, Deserialize)]
pub struct SearchQuery {
    q: String,
}

fn check_lang(lang: Option<&str>) -> Result<(), ApiError> {
    match lang {
        Some(value) if !LANGS.contains(&value) => Err(ApiError::invalid_field(
            "lang",
            "one_of",
            format!("must be one of {}", LANGS.join(", ")),
        )),
        _ => Ok(()),
    }
}

fn matches_text(article: &Article, needle: &str) -> bool {
    article.title.to_lowercase().contains(needle)
        || article.excerpt.to_lowercase().contains(needle)
}

#[get("/articles")]
async fn list_articles(
    state: web::Data<AppState>,
    query: web::Query<ListQuery>,
) -> Result<HttpResponse, ApiError> {
    let (page, limit) = page_limit(query.page, query.limit)?;
    check_lang(query.lang.as_deref())?;

    let store = state.read();

    // An unknown facet slug is an empty page, not a 404, so `Some(None)` short
    // circuits instead of erroring.
    let category = query.category.as_deref().map(|slug| store.category_id_of(slug));
    let tag = query.tag.as_deref().map(|slug| store.tag_id_of(slug));
    let author = query.author.as_deref().map(|slug| store.author_by_slug(slug).map(|a| a.id));
    let unknown_facet = category == Some(None) || tag == Some(None) || author == Some(None);

    let needle = query.q.as_deref().map(str::to_lowercase);

    let matched: Vec<&Article> = if unknown_facet {
        Vec::new()
    } else {
        store
            .listing()
            .filter(|article| {
                category.flatten().is_none_or(|id| article.category_id == id)
                    && tag.flatten().is_none_or(|id| article.tag_ids.contains(&id))
                    && author.flatten().is_none_or(|id| article.author_id == id)
                    && query.lang.as_deref().is_none_or(|lang| article.lang == lang)
                    && needle.as_deref().is_none_or(|needle| matches_text(article, needle))
            })
            .collect()
    };

    let total = matched.len();
    let items = matched
        .into_iter()
        .skip(offset(page, limit))
        .take(limit as usize)
        .map(|article| dto::summary(&store, article))
        .collect();

    Ok(HttpResponse::Ok().json(Page::new(items, page, limit, total)))
}

#[get("/articles/{slug}")]
async fn get_article(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let store = state.read();
    let article = store.article_by_slug(&path).ok_or(ApiError::NotFound("article"))?;
    Ok(HttpResponse::Ok().json(dto::detail(&store, article)))
}

#[get("/categories")]
async fn list_categories(state: web::Data<AppState>) -> HttpResponse {
    let store = state.read();
    let mut counts: HashMap<u32, usize> = HashMap::new();
    for article in store.listing() {
        *counts.entry(article.category_id).or_default() += 1;
    }
    let items: Vec<Facet<'_>> = store
        .categories
        .iter()
        .map(|category| Facet {
            id: category.id,
            slug: &category.slug,
            name: &category.name,
            article_count: counts.get(&category.id).copied().unwrap_or(0),
        })
        .collect();
    HttpResponse::Ok().json(items)
}

#[get("/tags")]
async fn list_tags(state: web::Data<AppState>) -> HttpResponse {
    let store = state.read();
    let mut counts: HashMap<u32, usize> = HashMap::new();
    for article in store.listing() {
        for id in &article.tag_ids {
            *counts.entry(*id).or_default() += 1;
        }
    }
    let items: Vec<Facet<'_>> = store
        .tags
        .iter()
        .map(|tag| Facet {
            id: tag.id,
            slug: &tag.slug,
            name: &tag.name,
            article_count: counts.get(&tag.id).copied().unwrap_or(0),
        })
        .collect();
    HttpResponse::Ok().json(items)
}

#[get("/authors/{slug}")]
async fn get_author(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let store = state.read();
    let author = store.author_by_slug(&path).ok_or(ApiError::NotFound("author"))?;
    let article_count = store.listing().filter(|a| a.author_id == author.id).count();
    Ok(HttpResponse::Ok().json(AuthorDetail {
        id: author.id,
        slug: &author.slug,
        name: &author.name,
        bio: &author.bio,
        article_count,
    }))
}

#[get("/companies")]
async fn list_companies(
    state: web::Data<AppState>,
    query: web::Query<CompanyQuery>,
) -> Result<HttpResponse, ApiError> {
    let (page, limit) = page_limit(query.page, query.limit)?;
    if let Some(stage) = query.stage.as_deref()
        && !STAGES.contains(&stage)
    {
        return Err(ApiError::invalid_field(
            "stage",
            "one_of",
            format!("must be one of {}", STAGES.join(", ")),
        ));
    }

    let store = state.read();
    let matched: Vec<&crate::store::Company> = store
        .companies
        .iter()
        .filter(|company| {
            query.industry.as_deref().is_none_or(|value| company.industry == value)
                && query.stage.as_deref().is_none_or(|value| company.stage == value)
                && query.min_funding.is_none_or(|min| company.total_funding_usd >= min)
        })
        .collect();

    let total = matched.len();
    let items =
        matched.into_iter().skip(offset(page, limit)).take(limit as usize).collect::<Vec<_>>();

    Ok(HttpResponse::Ok().json(Page::new(items, page, limit, total)))
}

#[get("/search")]
async fn search(
    state: web::Data<AppState>,
    query: web::Query<SearchQuery>,
) -> Result<HttpResponse, ApiError> {
    let length = query.q.chars().count();
    if !(2..=100).contains(&length) {
        return Err(ApiError::invalid_field("q", "length", "must be 2 to 100 characters"));
    }
    let needle = query.q.to_lowercase();
    let store = state.read();

    let articles = store
        .listing()
        .filter(|article| matches_text(article, &needle))
        .take(10)
        .map(|article| dto::summary(&store, article))
        .collect();

    let companies = store
        .companies
        .iter()
        .filter(|company| {
            company.name.to_lowercase().contains(&needle)
                || company.slug.contains(&needle)
                || company.industry.contains(&needle)
        })
        .take(10)
        .collect();

    Ok(HttpResponse::Ok().json(SearchResults { query: &query.q, articles, companies }))
}

pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(list_articles)
        .service(get_article)
        .service(list_categories)
        .service(list_tags)
        .service(get_author)
        .service(list_companies)
        .service(search);
}
